tests/test_fs_plugins.py

```diff
@@ -0,0 +1,181 @@
+        response = client.get_object(Bucket=bucket_name, Key=key_name)
+        print(bucket_name)
+        print(key_name)
+        return response["Body"].read()
+
+    def write_file(key_name, body):
+        client.put_object(Bucket=bucket_name, Key=key_name, Body=body)
+
+    return get_uri, read_file, write_file
+
+
+MAP_URI_HELPER = {S3_URI: s3_init}
+
+
+# ------------------------ URI CONDITION FOR EACH TEST ----------------------- #
+
+
+uri_re = re.compile(r"[\+\-][^:]+")
+
+
+def check_uri_skip_condition(uri, cond_str):
+    """Check whether a test should be skip or not.
+    Sometime, we want a test to run only with certain `uri`,
+    we could use `cond_str` to control which `uri` should be tested.
+
+    Args:
+        uri: The `uri` we want to check.
+        cond_str: A string that describes the `uri` condition for a test.
+            Every `uri` that matches the condition will be tested with that test.
+            A cond_str is in the form of "+uri:-uri". "+all" will add all uri
+            into the condition. Note that this parametter is parsed from left to right.
+            E.g: "+all:-s3" will run that test will all `uri` except `s3`.
+    """
+    cond_set = None
+    if cond_str not in check_uri_skip_condition.cached_cond:
+        cond_set = set()
+        for cond in uri_re.findall(cond_str):
+            sign, cond_uri = cond[0], cond[1:]
+            if sign == "+":
+                if cond_uri == "all":
+                    cond_set.update(ALL_URI)
+                else:
+                    cond_set.add(cond_uri)
+            else:
+                if cond_uri == "all":
+                    cond_set.clear()
+                else:
+                    cond_set.discard(cond_uri)
+        check_uri_skip_condition.cached_cond[cond_str] = cond_set
+    else:
+        cond_set = check_uri_skip_condition.cached_cond[cond_str]
+    if uri not in cond_set:
+        pytest.skip()
+
+
+check_uri_skip_condition.cached_cond = {}
+
+
+def check_test_condition_and_setup_env(uri, envs, monkeypatch):
+    uri_cond, additional_env = envs
+    check_uri_skip_condition(uri, uri_cond)
+    if isinstance(additional_env, dict):
+        for key, value in additional_env.items():
+            if value is not None:
+                monkeypatch.setenv(key, value)
+            else:
+                monkeypatch.delenv(key)
+
+
+@pytest.fixture(
+    scope="module",
+    autouse=True,
+    params=[
+        S3_URI,
+        pytest.param(AZ_URI, marks=pytest.mark.skip(reason="TODO")),
+        pytest.param(HDFS_URI, marks=pytest.mark.skip(reason="TODO")),
+        pytest.param(VIEWFS_URI, marks=pytest.mark.skip(reason="TODO")),
+        pytest.param(HAR_URI, marks=pytest.mark.skip(reason="TODO")),
+        pytest.param(GCS_URI, marks=pytest.mark.skip(reason="TODO")),
+    ],
+)
+def uri_init(request):
+    uri = request.param
+    monkeypatch = pytest.MonkeyPatch()
+    yield uri, *MAP_URI_HELPER[uri](monkeypatch)
+    monkeypatch.undo()
```

Comment:
Because the scope of this fixture is `module`, it will be called once for each `uri`. With [monkeypatch](https://docs.pytest.org/en/stable/monkeypatch.html), all the envs for each `uri` will be unset when we call `monkeypatch.undo()`. [`yield`](https://docs.pytest.org/en/stable/fixture.html#fixture-finalization-executing-teardown-code) is a technique allows us to setup and teardown fixture in one function ( and that where we need to call `monkeypatch.undo()` to unset the envs for that `uri` )

So one layer of isolation, env for a `uri` won't be clash with other `uri`.

Reply:
This sounds good to me 👍